compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -220,6 +220,12 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let mut vis = self.lower_visibility(&i.vis, None);
 
         if let ItemKind::MacroDef(MacroDef { ref body, macro_rules }) = i.kind {
+            if macro_rules {
```

Comment:
Should we keep this `if macro_rules` at all?

Reply:
Hm, good question.  So the other code path (for exported or 2.0 macros) calls `lower_attrs` which ends up stuffing the attributes in `Crate.attrs`, which will get visited for the `UnusedAttributes` lint.  For `macro_export`ed `macro_rules` items, that means those attributes will get visited twice by `UnusedAttributes`, but since it marks attributes used, the second visit will ignore them.

I don't know the codebase well enough to say if maybe `Crate.macro_attrs` and `Crate.attrs` could just be merged into one `Crate.attrs`.  That is, remove `macro_attrs` and call `lower_attrs` unconditionally.

Reply:
I'm not sure why `lower_item` needs to be changed.

All attributes should be available during late lint checking when the `unused_attributes` lint works, but it should visit both `exported_macros` and `non_exported_macro_attrs` instead of... whatever is done currently, I'm not sure how things work after #79519 anymore.

Reply:
Indeed, @petrochenkov is right, we do not need to change `lower_item`. All the attributes for exported macros are stored in `krate.attrs`. We just need to iterate on `non_exported_macro_attrs` in `LateLintPass::check_crate`.

Reply:
I pushed an update that changes it to just walk `non_exported_macro_attrs`.  I wasn't sure where to put it, since it is a bit of a maze to me.

It is passing in the wrong HIR ID since there is no HIR ID for those attributes.  Not sure what to do there. That's why I was asking is maybe it could `lower_attrs` so that all the attributes are in one place with HIR IDs (although for unexported macros, that HIR ID wouldn't match with anything in the HIR).

This change drops the `#[inline]` check for unexported macros, but it shows up as unused, so that's probably fine.

Let me know if this is what you were thinking.